File: cci/partition.go
package cci

/*
    Partition: partition a linked list around a value x, such that all nodes less than x come before all nodes
    greater than x.
*/

import "github.com/Ahmed-Sermani/algos/structures"

func Partition(head *structures.Node[int], pivot int) *structures.Node[int] {
    if head == nil {
        return head
    }
    var partition1, partition2, head1, head2 *structures.Node[int]
    
    curr := head
    for curr != nil {
        if curr.Data < pivot {
            if partition1 == nil {
                partition1 = &structures.Node[int]{Data: curr.Data}
                head1 = partition1
            } else {
                partition1.Next = &structures.Node[int]{Data: curr.Data}
                partition1 = partition1.Next
            }
        } else {
            if partition2 == nil {
                partition2 = &structures.Node[int]{Data: curr.Data}
                head2 = partition2
            } else {
                partition2.Next = &structures.Node[int]{Data: curr.Data}
                partition2 = partition2.Next
            }
        }
    }
    partition1.Next = head2
    *head = *head1
    return head
}

func Partition2(node *structures.Node[int], pivot int) *structures.Node[int] {
    head := node
    tail := node

    for node != nil {
        next := node.Next
        if node.Data < pivot {
            node.Next = head
            head = node
        } else {
            tail.Next = node
            tail = node
        }
        node = next
    }
    tail.Next = nil
    return head
}

File: python/ch2/partition.py



def partition(node, x):
    
    before_start = before_end = after_start = after_end = None

    while node is not None:
        next = node.next
        node.next = None
        if node.val < x:
            if before_start is None:
                before_start = node
                before_end = before_start
            else:
                before_end.next = node
                before_end = node
        else:
            if after_start is None:
                after_start = node
                after_end = after_start
            else:
                after_end.next = node
                after_end = node

        node = next

    if before_start is None:
        return after_start

    # merge
    before_end.next = after_start

    return before_start



def partition2(node, x):

    if node is None:
        return

    head = tail = node

    while node is not None:
        next = node.next
        if node.val < x:
            node.next = head
            head = node
        else:
            tail.next = node 
            tail = node

        node = next

    tail.next = None

    return head


